### DISCUSSION TITLE
Is there a way to access cluster metrics or stats ?

### DISCUSSION THREAD

**Discussion by Pier4413** [kube-rs/kube#1299] [Q&A]
As the title says. Is there a way to access the data in [Node metrics data](https://kubernetes.io/docs/reference/instrumentation/node-metrics/)

The exemple provided by [Node metrics data](https://kubernetes.io/docs/reference/instrumentation/node-metrics/) is for minikube but I could access it on other workers. I don't use minikube but another framework that create the same route except the minikube part of the uri is replaced by another worker name.

Currently I have added a reqwest call to my code but that bother me as this is a specific use case and my other functions use Kube.rs

I there a way to access data via something generic like that :

/api/v1/nodes/${WORKER_NAME}/proxy/stats/summary using Kube.rs ?

**Comment by mateiidavid** [comment]
Hey @Pier4413, I toyed with the following snippet to answer a different question on the same topic:

```rust
use kube::{core::DynamicObject, Api, Resource, ResourceExt};
use serde::Deserialize;

#[tokio::main]
async fn main() -> anyhow::Result<()> {
    let client = kube::Client::try_default().await?;
    let apigroup = kube::discovery::group(&client, "metrics.k8s.io").await?;
    let res = apigroup.recommended_resources();
    for (ar, _caps) in res {
        println!("Ar: {ar:?}");
    }
    let (ar, _caps) = apigroup.recommended_kind("NodeMetrics").unwrap();
    let api: Api<DynamicObject> = Api::all_with(client.clone(), &ar);
    for obj in api.list(&Default::default()).await? {
        let (name, metric) = extract_metric(obj)?;
        println!("{name}: {metric:?}");
    }
    Ok(())
}

#[derive(Default, Deserialize, Clone, Debug)]
struct Metric {
    cpu: String,
    memory: String,
}

fn extract_metric(obj: DynamicObject) -> anyhow::Result<(String, Metric)> {
    let metric = obj
        .data
        .get("usage")
        .cloned()
        .ok_or_else(|| anyhow::anyhow!("missing 'usage'"))?;
    let metric = serde_json::from_value(metric)?;
    Ok((obj.name_any(), metric))
}
```

This is a functional equivalent of `kubectl top <node>` and it uses a client to dynamically resolve the type of the metrics API resource at runtime (simply because I had no idea what the supported kinds and versions are ahead of time). This would require you to use [the metrics API server](https://github.com/kubernetes-sigs/metrics-server). 

I didn't find a way to dynamically fetch the timeseries data without manually creating an http client and hitting the `node/proxy` endpoint.  I looked through the [Go code](https://github.com/kubernetes/kubernetes/blob/7d309e0104fedb57280b261e5677d919cb2a0e2d/staging/src/k8s.io/kubelet/pkg/apis/stats/v1alpha1/BUILD) and it seems that `node/proxy/summary` doesn't have a concrete type. Probably because this is all fetched from the kubelet's admin endpoint. There doesn't seem to be a resource that allows us to build a client.

I'd either build an abstraction over the kubelet's admin endpoint or just use a timeseries ingester and use a provided client to query this data.
